src/lib/Libwin/accesinfo.c

```diff
@@ -852,7 +852,7 @@ perm_granted_admin_and_owner(char *path, int disallow, char *owner, char *errmsg
 			 ( EqualSid((SID *)&pace->SidStart, esid) || \
 			sid2rid((SID *)&pace->SidStart) != \
 					 SECURITY_CREATOR_OWNER_RID ) && \
-			(!sidIsAdminPrivilege( (SID *)&pace->SidStart)) && \
+			(!sidIsAdminPrivilege( (SID *)&pace->SidStart)) || \
```

Comment:
I believe in here, we add to the "allowed" list if file's owner has admin privilege, or if not admin-privilege, and yet there's a passed 'owner' to the function (i.e. owner != NULL), that owner must be the same as the file's owner (i.e. name). So I think the if condition here should be:
 && (sidIsAdminPrivilege( (SID *)&pace->SidStart) || ((owner != NULL) && (strcmp(name, owner) == 0))
*** kind of like the opposite of the one line 814 that we're saying is invalid.

Reply:
Thanks for your review @bayucan 

I am reading the code this way, 
- In the first check after entering this api at https://github.com/PBSPro/pbspro/blob/master/src/lib/Libwin/accesinfo.c#L814, we are validating whether the file is owned by "Admin" user or passed "owner", else throwing EPERM error(1).
-  If the control goes further means, file is owned by Admin/passed owner. then we are getting ACL and ACE's. 
- At second check, in loop, we are parsing each ACE and filtering only the allowed list of file permissions. here. we are interested only in ***Non***-Admin's, ***Non***-passed-owner's allowed file attributes. So that we check with passed mask value for NON-admins/NON-passed-owners permissions. 
Eg: if we dont want to allow NON-admin write permissions, we are filtering only non-admin users allowed permissions.
- Later we are filtering "Disallowed" permissions and comparing all data with disallow mask flag and return EACCES(13) error.

I have attached test results for below four cases in attached doxs with all possibilites, its working as expected without proposed changes.
1) Admin only owned file without sharing write permission to Non-admin users.
2) Admin owned, but sharing write permission with non-admin user
3) Non-admin owned file
4) Non-Admin owned file with sharing write permission with Non-admin users

FYI: I have tried building with suggested way, It seems not working in perfect case.(ie: Admin owned , Sharing no write permissions with NON-admin user.)

Kindly let me know if understaning need some change. Thanks

Reply:
Your explanation makes sense! Thanks @bremanandjk